src/supervisor.cr

```diff
@@ -3,24 +3,74 @@ require "wait_group"
 require "./agent"
 
 module Earl
-  # Supervises other agents.
+  # Supervise other agents.
+  #
+  # Use a `Supervisor` to monitor long lived agents that shall be spawned before
+  # the program starts and are expected to stay alive for the whole duration of
+  # the program, or at worst will stop by themselves after some time and won't
+  # ever need to be restarted.
   #
   # The list of monitored agents is fixed and all agents must be added to the
   # supervisor before the supervisor starts.
   #
-  # Use a `Supervisor` to monitor long lived agents that shall be spawned before
-  # the program starts and expectd expected to stay alive for the whole duration
-  # of the program, or at worst will stop by themselves after some time and
-  # won't ever need to be restarted.
+  # Each agent is spawned within its own dedicated `Fiber`. The supervisor takes
+  # care to recycle and restart crashed agents. The supervisor automatically
+  # shuts down when all its agents have stopped normally.
+  #
+  # ### Restart intensity
   #
-  # - Spawns agents in their dedicated `Fiber`.
-  # - Recycles and restarts crashed agents.
-  # - Eventually stops when all agents have stopped.
+  # To avoid an infinite loop when a monitored agent continuously crashes, a
+  # supervisor defines a maximum restart intensity, that is a maximum number of
+  # restart counts over a period of time. When the allowed intensity is reached,
+  # then the
+  #
+  # For example with an +intensity+ of 1 and a +period+ of `5.seconds` then an
+  # agent restarting twice within the last 5 seconds will cause the supervisor
+  # to shutdown. The same agent crashing once within those 5 seconds or
+  # repeatedly crashing every 6 seconds will always be restarted normally.
   class Supervisor
     include Agent
 
-    def initialize
-      @agents = [] of Agent
+    # :nodoc:
+    #
+    # Wraps an agent to keep additional metadata such as the number of restart
+    # and when they happened.
+    class Supervised
```

Comment:
Not very important because it's an internal API, but I'm not fond of the `Supervised` name. Maybe `Metadata`?